File: Set 8/Jump Game 2/Jump.py
def min_jumps(arr):
    n = len(arr)
    if n <= 1:
        return 0
    if arr[0] == 0:
        return -1

    jumps = 1
    max_reach = arr[0]
    steps = arr[0]

    for i in range(1, n):
        if i == n - 1:
            return jumps

        max_reach = max(max_reach, i + arr[i])
        steps -= 1

        if steps == 0:
            jumps += 1
            if i >= max_reach:
                return -1
            steps = max_reach - i

    return -1
    
    
t = int(input("Enter number of test cases: "))
for _ in range(t):
    arr = list(map(int, input("Enter array elements: ").split()))
    result = min_jumps(arr)
    if result == -1:
        print("Cannot reach the end of the array.")
    else:
        print("Minimum number of jumps:", result)

File: Set 8/Jump Game 2/Jump.cpp
#include <iostream>
#include <vector>
#include <climits>
using namespace std;

int minJumps(vector<int>& arr) {
    int n = arr.size();
    if (n <= 1) return 0;
    if (arr[0] == 0) return -1;

    int jumps = 1;
    int maxReach = arr[0];
    int steps = arr[0];

    for (int i = 1; i < n; i++) {
        if (i == n - 1) return jumps;

        maxReach = max(maxReach, i + arr[i]);
        steps--;

        if (steps == 0) {
            jumps++;
            if (i >= maxReach) return -1;
            steps = maxReach - i;
        }
    }
    return -1;
}

int main() {
    int t;
    cout << "Enter number of test cases: ";
    cin >> t;

    while (t--) {
        int n;
        cout << "Enter the size of the array: ";
        cin >> n;

        vector<int> arr(n);
        cout << "Enter the array elements: ";
        for (int i = 0; i < n; i++) {
            cin >> arr[i];
        }

        int result = minJumps(arr);
        if (result == -1)
            cout << "Cannot reach the end of the array." << endl;
        else
            cout << "Minimum number of jumps: " << result << endl;
    }

    return 0;
}
